### PR TITLE
Update Json middleware benchmark to use source-gen

### PR DESCRIPTION
**Pull request by BrennanConroy**
179m bytes per second to 3m
Working Set 440MB to 62MB
RPS change unknown right now since the baseline changed between running earlier in the day and now. But running them back-to-back right now gives a ~1.1% increase in RPS

### CODE REVIEW

File: src/Benchmarks/Middleware/JsonMiddleware.cs [hunk 1]
```diff
@@ -53,7 +56,11 @@ public Task Invoke(HttpContext httpContext)
 
                 return Task.CompletedTask;
 #else
-                return JsonSerializer.SerializeAsync<JsonMessage>(httpContext.Response.Body, new JsonMessage { message = "Hello, World!" });
+                _writer ??= new Utf8JsonWriter(httpContext.Response.Body);
+                _writer.Reset(httpContext.Response.Body);
+
+                JsonSerializer.Serialize<JsonMessage>(_writer, new JsonMessage { message = "Hello, World!" }, CustomJsonContext.Default.JsonMessage);
```
Comment (DamianEdwards) [inline src/Benchmarks/Middleware/JsonMiddleware.cs:62]: I thought `JsonSerializer.Serialize` will internally pool the `Utf8JsonWriter` instance it uses if we don't pass one?

Comment (BrennanConroy) [reply src/Benchmarks/Middleware/JsonMiddleware.cs:62]: No https://source.dot.net/#System.Text.Json/System/Text/Json/Serialization/JsonSerializer.Write.Stream.cs,391
And also, the source-gen isn't called when using the stream overloads. https://github.com/dotnet/runtime/pull/78646

Comment (DamianEdwards) [reply src/Benchmarks/Middleware/JsonMiddleware.cs:62]: So which overloads do pool? Just those that return string?

Comment (BrennanConroy) [reply src/Benchmarks/Middleware/JsonMiddleware.cs:62]: > Just those that return string?

Looks like it

File: src/Benchmarks/Middleware/JsonMiddleware.cs [hunk 2]
```diff
@@ -10,7 +10,7 @@
 using Microsoft.AspNetCore.Http;
 using Microsoft.AspNetCore.Http.Features;
 
-#if NETCOREAPP3_0 || NETCOREAPP3_1 || NETCOREAPP5_0 || NET5_0 || NET6_0_OR_GREATER
+#if NETCOREAPP3_0_OR_GREATER
```
[the remark this answers is not in the document]
Comment (BrennanConroy) [reply src/Benchmarks/Middleware/JsonMiddleware.cs:13]: @sebastienros ?

[the remark this answers is not in the document]
Comment (sebastienros) [reply src/Benchmarks/Middleware/JsonMiddleware.cs:13]: I changed other benchmarks recently to target 6 at mininum. I think we can do the same here as we haven't had to run it even for 5 since last year.

The excuse to not do it earlier is that we always talk about getting rid of this project and refactor it in more granular apps (middleware, mvc) and updating it to the latest templates. That some good amount of work because we will need to update the benchmarks configurations and CI too. But this could be done in multiple steps.

File: src/Benchmarks/Middleware/JsonMiddleware.cs [hunk 3]
```diff
@@ -40,11 +37,8 @@ public Task Invoke(HttpContext httpContext)
                     syncIOFeature.AllowSynchronousIO = true;
                 }
 
-                _writer ??= new Utf8JsonWriter(httpContext.Response.Body);
-                _writer.Reset(httpContext.Response.Body);
-
-                JsonSerializer.Serialize<JsonMessage>(_writer, new JsonMessage { message = "Hello, World!" }, CustomJsonContext.Default.JsonMessage);
-                return Task.CompletedTask;
+                return JsonSerializer.SerializeAsync<JsonMessage>(httpContext.Response.Body, new JsonMessage { message = "Hello, World!" },
```
Comment (sebastienros) [inline src/Benchmarks/Middleware/JsonMiddleware.cs:40]: Remove `.AllowSynchronousIO = true` ?

Comment (BrennanConroy) [reply src/Benchmarks/Middleware/JsonMiddleware.cs:40]: Maybe, forgot pushing to the branch would update a PR 😄 
Still need to test and see if the recent STJ change did anything

Comment (BrennanConroy) [reply src/Benchmarks/Middleware/JsonMiddleware.cs:40]: @eiriktsarpalis, I thought https://github.com/dotnet/runtime/pull/78646 was supposed to allow source-gen code to be used with `SerializeAsync`? But when trying I get the exception shown in https://github.com/dotnet/runtime/issues/71933, without creating a combined context. Are we using it wrong here, or is this another version of the same bug and source-gen still isn't usable with `SerializeAsync`?

Comment (BrennanConroy) [reply src/Benchmarks/Middleware/JsonMiddleware.cs:40]: Removing the `GenerationMode = JsonSourceGenerationMode.Serialization` argument made it work. Is that required for `SerializeAsync`?

**Comment by BrennanConroy** [comment]
176 MB/sec -> 33 MB/sec with Source gen + new fast path streaming serialization.
Also got 1.006m RPS, the benchmark averages ~988k so slight improvement.

File: src/Benchmarks/Middleware/JsonMiddleware.cs (hunk 3 above)
Comment (eiriktsarpalis) [reply src/Benchmarks/Middleware/JsonMiddleware.cs:40]: > Removing the GenerationMode = JsonSourceGenerationMode.Serialization argument made it work. Is that required for SerializeAsync?

Streaming serialization requires metadata generation. The change in dotnet/runtime#78646 simply introduces a heuristic that eventually falls back to fast-path serialization provided that the JSON payloads are small enough.

File: src/Benchmarks/Middleware/JsonMiddleware.cs [hunk 4]
```diff
@@ -66,6 +50,12 @@ public static class JsonMiddlewareExtensions
         public static IApplicationBuilder UseJson(this IApplicationBuilder builder) => builder.UseMiddleware<JsonMiddleware>();
     }
 
+    [JsonSourceGenerationOptions]
```
Comment (eiriktsarpalis) [inline src/Benchmarks/Middleware/JsonMiddleware.cs:53]: Can be removed at this point:

```suggestion
```

File: src/Benchmarks/Middleware/JsonMiddleware.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/Benchmarks/Middleware/JsonMiddleware.cs]: All serialization overloads pool `Utf8JsonWriter`, with the exception of the async streaming serialization overloads that can't use the fast path.

File: src/Benchmarks/Middleware/JsonMiddleware.cs (hunk 3 above)
Comment (DamianEdwards) [reply src/Benchmarks/Middleware/JsonMiddleware.cs:40]: @BrennanConroy any reason we can't simply use the `WriteAsJsonAsync` extension methods here, passing in the options instance from DI resolved from the middleware constructor (to avoid request services)? I assume you'd have to add the serialization context to the options during DI setup too. Would make this slightly more "real world" as it'd use JSON web defaults then.

Comment (BrennanConroy) [reply src/Benchmarks/Middleware/JsonMiddleware.cs:40]: Took forever, but I got it working. One unfortunate thing you have to do to make it work using `WriteAsJsonAsync` is to set `JsonSerialierOptions.Encoder` to `null`.

Comment (DamianEdwards) [reply src/Benchmarks/Middleware/JsonMiddleware.cs:40]: You going to push the update so we can see? :smile:

Comment (BrennanConroy) [reply src/Benchmarks/Middleware/JsonMiddleware.cs]: Enjoy

File: src/Benchmarks/Program.cs [hunk 5]
```diff
@@ -84,6 +85,11 @@ public static void Main(string[] args)
                         }
 #endif
                     })
+                    .ConfigureHttpJsonOptions(jsonOptions =>
+                    {
+                        jsonOptions.SerializerOptions.Encoder = null;
```
Comment (eerhardt) [inline src/Benchmarks/Program.cs:90]: Why do we need this? It seems like a bug.... At least a comment explaining why.

Comment (BrennanConroy) [reply src/Benchmarks/Program.cs:90]: https://source.dot.net/#System.Text.Json/System/Text/Json/Serialization/JsonSerializerContext.cs,60
The fast path logic requires very specific settings. Didn't add a comment because not sure we want to use this approach in the benchmark.

File: src/Benchmarks/Middleware/JsonMiddleware.cs [hunk 6]
```diff
@@ -50,7 +49,7 @@ public static class JsonMiddlewareExtensions
         public static IApplicationBuilder UseJson(this IApplicationBuilder builder) => builder.UseMiddleware<JsonMiddleware>();
     }
 
-    [JsonSourceGenerationOptions]
+    [JsonSourceGenerationOptions(PropertyNamingPolicy = JsonKnownNamingPolicy.CamelCase)]
```
Comment (DamianEdwards) [inline src/Benchmarks/Middleware/JsonMiddleware.cs:52]: Why is this needed if we're using the instance of the serializer options from DI that's already configured to use this casing by default?

File: src/Benchmarks/Middleware/JsonMiddleware.cs [hunk 7]
```diff
@@ -20,25 +22,22 @@ public class JsonMiddleware
...
 
-                var syncIOFeature = httpContext.Features.Get<IHttpBodyControlFeature>();
-                if (syncIOFeature != null)
-                {
-                    syncIOFeature.AllowSynchronousIO = true;
-                }
-
-                return JsonSerializer.SerializeAsync<JsonMessage>(httpContext.Response.Body, new JsonMessage { message = "Hello, World!" },
-                    CustomJsonContext.Default.JsonMessage);
+                return httpContext.Response.WriteAsJsonAsync<JsonMessage>(new JsonMessage { message = "Hello, World!" }, _jsonOptions, httpContext.RequestAborted);
```
Comment (DamianEdwards) [inline src/Benchmarks/Middleware/JsonMiddleware.cs:40]: You don't need to specify the type arg here do you?

File: src/Benchmarks/Middleware/JsonMiddleware.cs (hunk 6 above)
Comment (BrennanConroy) [reply src/Benchmarks/Middleware/JsonMiddleware.cs:52]: Because we add this context to the options, and when it serializes JsonMessage this context is used which has the default of `null` naming policy. And the naming policy needs to match the main options in https://source.dot.net/#System.Text.Json/System/Text/Json/Serialization/JsonSerializerContext.cs,74

I of course could be doing something wrong, these Json APIs are way too complex.

File: src/Benchmarks/Program.cs [hunk 8]
```diff
@@ -84,6 +85,11 @@ public static void Main(string[] args)
...
+                        jsonOptions.SerializerOptions.AddContext<Middleware.CustomJsonContext>();
```
Comment (eerhardt) [inline src/Benchmarks/Program.cs:91]: 1. AddContext is going to be obsoleted as per https://github.com/dotnet/runtime/issues/83280.
2. I'm a bit confused at how this works today. `AddContext` [will put the `CustomJsonContext` at the end of the TypeInfoResolver](https://github.com/dotnet/runtime/blob/967250cc86b4c613015066b47b882e7424fdb590/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs#L175-L179), after the "default" resolver. Doesn't this mean the reflection based type info resolver still gets used over the `CustomJsonContext`?

cc @eiriktsarpalis

Comment (BrennanConroy) [reply src/Benchmarks/Program.cs:91]: This worked temporarily until AddContext was fixed to properly combine contexts. I think my local branch fixed that but never pushed it.

File: src/Benchmarks/Middleware/JsonMiddleware.cs [hunk 9]
```diff
@@ -4,34 +4,38 @@
...
+using Microsoft.Extensions.Options;
+using Microsoft.AspNetCore.Http.Json;
 
 namespace Benchmarks.Middleware
 {
     public class JsonMiddleware
     {
         private static readonly PathString _path = new PathString(Scenarios.GetPath(s => s.Json));
-        private static readonly UTF8Encoding _encoding = new UTF8Encoding(false);
         private const int _bufferSize = 27;
```
Comment (davidfowl) [inline src/Benchmarks/Middleware/JsonMiddleware.cs:21]: How much performance do we lose if we remove this?

File: src/Benchmarks/Middleware/JsonMiddleware.cs [hunk 10]
```diff
@@ -4,34 +4,38 @@
...
         public Task Invoke(HttpContext httpContext)
         {
             if (httpContext.Request.Path.StartsWithSegments(_path, StringComparison.Ordinal))
             {
                 httpContext.Response.StatusCode = 200;
-                httpContext.Response.ContentType = "application/json";
                 httpContext.Response.ContentLength = _bufferSize;
 
-                return JsonSerializer.SerializeAsync<JsonMessage>(httpContext.Response.Body, new JsonMessage { message = "Hello, World!" });
+                return httpContext.Response.WriteAsJsonAsync(new JsonMessage { message = "Hello, World!" }, _jsonOptions, httpContext.RequestAborted);
```
Comment (davidfowl) [inline src/Benchmarks/Middleware/JsonMiddleware.cs:38]: Why are we passing the token?

File: src/Benchmarks/Middleware/JsonMiddleware.cs (hunk 6 above)
Comment (eiriktsarpalis) [reply src/Benchmarks/Middleware/JsonMiddleware.cs:50]: It is not strictly necessary, since serialization will ultimately conform to whatever configuration is passed by the relevant `JsonSerializerOptions`. The benefit of declaring here though is that any naming policies will be computed at compile time and incorporated in the fast-path method.

File: src/Benchmarks/Middleware/JsonMiddleware.cs [hunk 11]
```diff
@@ -43,6 +47,12 @@ public static class JsonMiddlewareExtensions
         public static IApplicationBuilder UseJson(this IApplicationBuilder builder) => builder.UseMiddleware<JsonMiddleware>();
     }
 
+    [JsonSourceGenerationOptions(PropertyNamingPolicy = JsonKnownNamingPolicy.CamelCase)]
```
Comment (eiriktsarpalis) [inline src/Benchmarks/Middleware/JsonMiddleware.cs:50]: It should now be possible to do this:
```suggestion
    [JsonSourceGenerationOptions(JsonSerializerDefaults.Web)]
```

File: src/Benchmarks/Program.cs (hunk 8 above)
Comment (eiriktsarpalis) [reply src/Benchmarks/Program.cs:87]: You should use this to ensure equivalent behavior with the .NET 8 build
```suggestion
                        jsonOptions.SerializerOptions.TypeInfoResolver = JsonTypeInfoResolver.Combine(jsonOptions.SerializerOptions.TypeInfoResolver, Middleware.CustomJsonContext.Default);
```

File: src/Benchmarks/Middleware/JsonMiddleware.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (DamianEdwards) [reply src/Benchmarks/Middleware/JsonMiddleware.cs:38]: Ideally you'd set it up in DI and not have to pass it at all here. Is that still not possible?

[the remark this answers is not in the document]
Comment (BrennanConroy) [reply src/Benchmarks/Middleware/JsonMiddleware.cs:38]: It should work with DI (you can see we get the options via DI in the ctor), but that's just extra per request work that we can amortize via the ctor.

[the remark this answers is not in the document]
Comment (eerhardt) [reply src/Benchmarks/Middleware/JsonMiddleware.cs:38]: Passing in the options, and making the serializer look up the JsonTypeInfo every request, is just extra work as well. If you just pass the JsonTypeInfo directly in, it is less work to do per request.

File: src/Benchmarks/Program.cs (hunk 8 above)
Comment (BrennanConroy) [reply src/Benchmarks/Program.cs:87]: Shouldn't it be `Combine(Middleware.CustomJsonContext.Default, jsonOptions.SerializerOptions.TypeInfoResolver);`? Or does order not matter with `Combine`?

File: src/Benchmarks/Middleware/JsonMiddleware.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (BrennanConroy) [reply src/Benchmarks/Middleware/JsonMiddleware.cs:38]: Yep, I'll try that one out

File: src/Benchmarks/Program.cs (hunk 8 above)
Comment (eiriktsarpalis) [reply src/Benchmarks/Program.cs]: You're right, I messed up the order in my suggestion.

File: src/Benchmarks/Middleware/JsonMiddleware.cs [hunk 12]
```diff
@@ -4,34 +4,53 @@
...
+#else
+        private readonly JsonSerializerOptions _jsonOptions;
+#endif
 
-        public JsonMiddleware(RequestDelegate next) => _next = next;
+        public JsonMiddleware(RequestDelegate next, IOptions<JsonOptions> jsonOptions)
+        {
+            _next = next;
+#if NET8_0_OR_GREATER
+            _jsonTypeInfo = JsonTypeInfo.CreateJsonTypeInfo<JsonMessage>(jsonOptions.Value.SerializerOptions);
```
Comment (BrennanConroy) [inline src/Benchmarks/Middleware/JsonMiddleware.cs:34]: @eiriktsarpalis is `JsonTypeInfo.CreateJsonTypeInfo<JsonMessage>(jsonOptions.Value.SerializerOptions);` supposed to create a `JsonTypeInfo<T>` that doesn't seem to work?

File: src/Benchmarks/Middleware/JsonMiddleware.cs (hunk 9 above)
Comment (BrennanConroy) [reply src/Benchmarks/Middleware/JsonMiddleware.cs]: We lose 20-30k RPS

File: src/Benchmarks/Middleware/JsonMiddleware.cs (hunk 12 above)
Comment (eiriktsarpalis) [reply src/Benchmarks/Middleware/JsonMiddleware.cs]: That just creates a blank, mutable `JsonTypeInfo` for the given type. You should use `(JsonTypeInfo<JsonMessage>)jsonOptions.Value.SerializerOptions.GetTypeInfo(typeof(JsonMessage));` instead.

File: src/Benchmarks/Middleware/JsonMiddleware.cs [hunk 13]
```diff
@@ -4,34 +4,51 @@
...
+#else
+        private readonly JsonSerializerOptions _jsonOptions;
+#endif
 
-        public JsonMiddleware(RequestDelegate next) => _next = next;
+        public JsonMiddleware(RequestDelegate next, IOptions<JsonOptions> jsonOptions)
+        {
+            _next = next;
+#if NET8_0_OR_GREATER
+            _jsonTypeInfo = jsonOptions.Value.SerializerOptions.GetTypeInfo(typeof(JsonMessage)) as JsonTypeInfo<JsonMessage>;
```
Comment (eiriktsarpalis) [inline src/Benchmarks/Middleware/JsonMiddleware.cs:33]: The generic is guaranteed to match the type you're passing it:
```suggestion
            _jsonTypeInfo = (JsonTypeInfo<JsonMessage>)jsonOptions.Value.SerializerOptions.GetTypeInfo(typeof(JsonMessage));
```

File: src/Benchmarks/Middleware/JsonMiddleware.cs [hunk 14]
```diff
@@ -4,34 +4,53 @@
...
                 httpContext.Response.ContentLength = _bufferSize;
 
-                return JsonSerializer.SerializeAsync<JsonMessage>(httpContext.Response.Body, new JsonMessage { message = "Hello, World!" });
+                return httpContext.Response.WriteAsJsonAsync(new JsonMessage { message = "Hello, World!" },
+#if NET8_0_OR_GREATER
+                    _jsonTypeInfo
+#else
+                    _jsonOptions
+#endif
+                    );
```
Comment (eerhardt) [inline src/Benchmarks/Middleware/JsonMiddleware.cs:53]: I think this is too complicated. It would make a ton of these changes much simpler if this was just:

```suggestion
                return httpContext.Response.WriteAsJsonAsync(new JsonMessage { message = "Hello, World!" }, CustomJsonContext.Default.JsonMessage);
```

Then get rid of the constructor changes, and the `ConfigureHttpJsonOptions` call.

File: src/Benchmarks/Program.cs (hunk 5 above)
Comment (eerhardt) [reply src/Benchmarks/Program.cs:84]: Is this still needed now that we aren't using the JsonOptions anymore?

File: src/Benchmarks/Benchmarks.csproj [hunk 15]
```diff
@@ -1,4 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk.Web">
+﻿<Project Sdk="Microsoft.NET.Sdk.Web">
```
Comment (eerhardt) [inline src/Benchmarks/Benchmarks.csproj:1]: (nit) can revert this file.

**Review by eerhardt — APPROVED** [review]
This looks great to me.

Do we have before and after numbers?

File: src/Benchmarks/Program.cs (hunk 5 above)
Comment (BrennanConroy) [reply src/Benchmarks/Program.cs:84]: Oh wow, we can remove it!

Comment (eerhardt) [reply src/Benchmarks/Program.cs]: To confirm we are using "fast path" on both .NET 8 AND 7 with the latest changes, right? Just making sure we are comparing apples-to-apples if we compare them against each other.
